### ISSUE TITLE
Optimise Ghost boot-up

### ISSUE THREAD

**Issue by ErisDS** [TryGhost/Ghost#2182]
This is a bit of a case of me thinking out loud. Perhaps it's part of #1789, but it is also related to #2156 and #2179 (package.json support).

At the moment we have:
- bootstrap (initialise the config file)
- server init (does a whole tonne of stuff, finally calling server.listen)
- serve first request = middleware

I'm wondering if the server init could be streamlined into only things that are absolutely necessary to happen before Ghost even starts.

Surely a whole heap of this isn't actually needed until we serve the first request which needs it. For example, mail - it might never ever be needed, so why bother initialising it before we start Ghost?

Similarly with the new package.json stuff, should we be loading this into config? That's certainly convenient, but should it be somewhere else, loaded only if and when it is necessary? Or should it be passed to config later on in the flow... generated when requested?

This is a pretty different way of working, but I'm thinking it could lead us to much more streamlined, more modular code, it could give us a faster boot time, and a smaller memory footprint by default. 

But I am just thinking out loud :wink:

**Comment by ErisDS** [comment]
@kirrg001 maybe an idea to look at this as part of the work to try to minimise the boot time.

Maybe there are things we can do to improve the structure. If not - it would be good to close this, it's pretty old.

**Comment by kirrg001** [comment]
Thanks for pointing me to this issue, wasn't aware of this one 🙈

When i have debugged the Ghost bootstrap last week, the most slow down parts where:

- we have required all kinds of bigger modules, which eat time to read them from disk, but they weren't needed on bootstrap
- lot's of sub dependencies had strict dependencies e.g. lodash, debug etc - this caused node to load a dependency multiple times, because e.g. lodash was installed 5 times with a different version (this has eaten bootstrap time and memory usage)
- all the other init calls where pretty fast (i have used `time-require` to make the whole bootstrap process visible)
- database connection eats lot's of time. one database connection is created by Ghost and one by knex-migrator to check the health of the database (this could be optimised by passing a database connection into knex-migrator) - will raise this as separate issue.

There are for sure pieces which we don't have to load on bootstrap, but i assume they won't bring a big boost. Closing for now.
